### PR TITLE
posix: Add iqueue submodule

### PR DESCRIPTION
**Pull request by jain-surabhi-23**
## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Passes tests (included warning check)
* [ ] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [ ] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by wesbland** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by minsii** [comment]
test:jenkins/ch4/ofi

**Comment by wesbland** [comment]
I just pushed a couple of patches in here to update a few things that we've changed in our version. It's mostly copyrights, but I also added a CVAR to tweak the size of the cells to make benchmarking easier.

For some reason, I can't see the updates on GitHub here, but I can see them on the command line.

**Comment by wesbland** [comment]
test:jenkins/ch4/ofi

**Comment by hzhou** [comment]
Could someone add some analyses on the motivations/benefits/benchmarks of this PR?

**Comment by minsii** [comment]
@hzhou We plan to use `iqueue` as the default protocol for shm/posix. I agree that the PR header should be updated with more information.

@wesbland Do you have any short summary and benchmark data that can share with us? I

**Comment by wesbland** [comment]
> @wesbland Do you have any short summary and benchmark data that can share with us? I

This shared memory module is similar to the code in CH3. It has an array of cells (called terminals) that are used for message transmission. When a process wants to send a message, it finds the next available cell for the receiving process. The main benefit here over fastboxes is that there are multiple cells available for a receiving process, which means that a large message or a series of small messages can be sent at the same time. The cells are not reserved for a particular process pair, but for any incoming messages.

For performance, iqueue should be the same or better in almost all cases except for small numbers of small messages.

File: src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The number of cells used for the depth of the iqueue.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_eager_init(int rank, int size)
+{
+    int mpi_errno = MPI_SUCCESS;
```
Comment (minsii) [inline src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: @raffenet @yfguo @hzhou fbox_init.h and iqueue_init.h are still header files. Do we plan to replace them with .c?

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: I started a PR to uninline fastbox but it proved more complicated than expected https://github.com/pmodels/mpich/pull/4209

Comment (yfguo) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: I prefer making it a .c file rather than .h.

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: Agree. Let's change it in this PR. @gcongiu It will be good if we first fix the issue in fbox. I am checking that PR now.

Comment (hzhou) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: I still have reservation on the contention concerns when multiple processes sending to a single process -- a very common application pattern. We need decide whether to abandon fastbox or workout a dynamic scheme that use both fastbox and iqueue at runtime before move forward with this PR.

Comment (hzhou) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: Then senders and receiver also contend on the queue, right?

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: @hzhou Yes, a `hybrid` (or dynamic) version that combines both fbox and iqueue would be ideal (that is what ch3 does). But we do not have time to work on it. And in fact the difference between the `hybrid` version and the `iqueue-only` version is very small (i.e., iqueue has worse small message latency). @pavanbalaji actually pointed out that there are some applications care about **small message latency**, but I am not sure how much difference we could see in application level. 

In summary, replacing `fbox` with `iqueue` will definitely give us benefit, and @gcongiu can go ahead with his GPU work which relies on `iqueue` (this is why we are pushing this PR now). We need more time to investigate the benefit of the `hybrid` version and I'd suggest we put its development in future plan. We can create a ticket to track this issue.

Comment (hzhou) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: How does GPU work depends on iqueue?

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: An example is that when IPC is not available, the sender has to copy the data from GPU to a temporary host buffer. A cell buffer in iqueue is a good option. The data transfer of an intranode send-receive can be implemented as:
gpu_buf on P0 -> cell buffer -> gpu_buf on P1

If we do the above thing with `fbox`, the performance will be likely always worse than that with `iqueue`. E.g., when chunking a large message into multiple pieces in order to fit into the fbox buffer.

As we have a milestone for GPU work, we need to prioritize the tasks. The most effective & less time option is to use iqueue as the default, and add the GPU functionality on top of it. @gcongiu please feel free to comment if I miss anything.

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: Correct, we always planned to use shared memory for intranode fallback when GPUDirect P2P is not usable thus, as Min said, the easiest way of doing this is through iqueues. As @pavanbalaji also noted this might be even faster than using IPC as memory handles have to be created to exchange addresses among processes before actual copy of data can start. Since creating these handle is very expensive and we don't know the location of the user buffer beforehand, iqueues might provide an effective intranode communication for the general case, not only fallback.

Comment (hzhou) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: > If we do the above thing with `fbox`, the performance will be likely always worse than that with `iqueue`. E.g., when chunking a large message into multiple pieces in order to fit into the fbox buffer.

You are chunking a large message into multiple pieces with cells too, right? The argument of performance is the same argument for general large messages, right? I am not saying it is wrong, but it seems to be an orthogonal argument.

There are benefits If we can keep the lower-layer transport orthogonal to the upper layer. Certainly one transport may perform better than another, but that is different from making GPU depend on one particular transport, as that will constrain the code in both layers.

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.h]: The GPU intranode fallback does not depend on any transport. It just depends on shared memory. The choice of iqueue is just for better performance. This is my understanding.

**Comment by minsii** [comment]
test:jenkins/ch4/ofi

**Comment by minsii** [comment]
Just notice that I am a co-author of the first (biggest) commit `ch4/shm: Add iqueue submodule`...I am also going to do some code cleanup for it. Can @gcongiu do a final review for it after I finish cleanup?

**Comment by minsii** [comment]
I pushed two minor updates. The iqueue commit is well cleaned and contain detailed comments. But I have concerns about the per-receiver terminal design, similar to @hzhou 's feedback. I put them as TODO comments in my second commit.  Ideally we want to study these options and address them before merging. But as we have other higher priority tasks, I'd suggest we put these TODO comments there and address them in future. We should open an ticket for it (I will do it).

**Comment by hzhou** [comment]
Conceptually, making dynamic choice using both fbox and iqueue is simple, right? Say, in `POSIX_eager_send`, check if fastbox is available and data fits, if so, call fbox_send, otherwise, call iqueue_send, if fail, enque_postponed. Did I miss anything?

**Comment by hzhou** [comment]
I guess the complication is the recv-side. The ordering maybe messy.

**Comment by gcongiu** [comment]
> All commits look ready to merge. @gcongiu could you please review the three commits, of which I am a co-author?
> 
> * ch4/iqueue: add TODO comments for several perf issues
> * ch4/iqueue: use more compact MPIR_ERR_CHECK
> * ch4/shm: Add iqueue submodule
> 
> @gcongiu : could you please also rename "HACK!! CHANGE DEFAULT FOR TESTING" when pushing the changes? I think the plan is to use iqueue as the default, so not HACK anymore.

Should I squash all the commits onto the first one?

**Comment by minsii** [comment]
@hzhou The only reason we do not further improve the code is because we do not have time to work on it. I think it is a reasonable shape now and we can put more optimizations in later.

**Comment by minsii** [comment]
@gcongiu No, I think separate commits are fine as long as we use the merge mode.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
Linking issue #3525.

**Comment by minsii** [comment]
Updated #3525 to cover the latest plan for SHM/POSIX active message.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@minsii when tests are complete, assuming they complete successfully, can you review the last commit?

**Comment by minsii** [comment]
@gcongiu Approved.
